### PR TITLE
LPD-20287 - Apply Data Set default visualization mode in the fragment

### PR DESCRIPTION
**Pull request by juanjofgliferay**
## Story / Task
[LPD-15879](https://liferay.atlassian.net/browse/LPD-15879) / [LPD-20287](https://liferay.atlassian.net/browse/LPD-20287)
Follow up for https://github.com/brianchandotcom/liferay-portal/pull/148367#issuecomment-2011250707

## Goal
Once an user has created a Data Set and has configured the visualization modes, we need to apply the configuration in the visualization of the Data Set in the fragment.
Consider:
- Data Set has no visualization mode configured - use table visualization
- Data Set has one visualization mode configured (no marked as default) - use configured visualization mode
- Data Set has more than one visualization mode configured (no one marked as default) - use first configured visualization mode found, using alphabetical order
- Data Set has more than one visualization mode configured (one of them marked as default) - use the visualization mode set as default

## UI
DSM configuration
image

FDS application
image

## Playwright
Screenshot from 2024-03-21 12-57-24

### Notes about tests
Made changes in the `fdsFragmentPage` to provide more robust checks when we need to test Drag'n Drop the Data Set fragment into a new page.

Running tests several times to ensure consistent success.
Screenshot from 2024-03-21 12-41-42

## Test Cases matrix
https://docs.google.com/spreadsheets/d/1GdTSw3KrMDtjEfpkDrih8JcWOTn22ZuIL8s8WPBDAyM/edit#gid=0

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js [hunk 1]
```diff
@@ -896,7 +896,11 @@ const FrontendDataSet = ({
 						/>
 					)}
 
-					<div className="data-set-wrapper" ref={wrapperRef}>
+					<div
+						className="data-set-wrapper"
+						data-testid={`visualization-mode-${activeView.name}`}
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js:901]: Use `data-testid` to reference the visualization mode as it main use is testing.

File: modules/test/playwright/tests/frontend-data-set-views-web/pages/FDSFragmentPage.ts [hunk 2]
```diff
@@ -29,6 +38,62 @@ export class FDSFragmentPage {
 		await this.page.goto('/');
 	}
 
+	async configureDataSetFragment({
```
Comment (juanjofgliferay) [inline modules/test/playwright/tests/frontend-data-set-views-web/pages/FDSFragmentPage.ts:41]: It is a wrapper of the steps needed to drag and drop the fragment.
My idea is to move them to private methods once the refactor of the `itemActions.spec` and `creationActions.spec` have been refactored to fix flakyness.

File: modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts [hunk 3]
```diff
@@ -222,12 +187,442 @@ test.describe('Data Set Settings', () => {
...
-			await test.step('Clean data', async () => {
-				await dataSetManagerApiHelpers.deleteDataSet({
-					erc: customDataSetERC,
-				});
-			});
 		});
 	});
 });
+
+export const fragmentTest = mergeTests(
```
Comment (juanjofgliferay) [inline modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts:194]: I don't like having mixed tests for DSM and fragment in the same file, even if they usually test how the configuration of one is applied in the other.

With the growth of the API that allow us to create all needed data without using the UI we could split tests based on its usage, x.e. adding a prefix for the same functionality, like `dsmSettings.spec.ts` and `fragmentSettings.spec.ts`.

Any thoughts?

File: modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts [hunk 4]
```diff
@@ -222,12 +187,442 @@ test.describe('Data Set Settings', () => {
...
+					const pageLayout =
+						await apiHelpers.headlessDelivery.createSitePage(
+							site.id,
+							getRandomString()
+						);
+
+					return pageLayout;
+				}
+			);
+
```
Comment (juanjofgliferay) [inline modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts:232]: I've tried to create a page including the Data Set fragment using:

```
const fdsFragmentDefinition = getFragmentDefinition(
	getRandomString(),
	'com.liferay.frontend.data.set.views.web.internal.fragment.renderer.FDSViewFragmentRenderer'
);

const layout = await fragmentTest.step('Create a new page', async () => {
	const pageLayout = await apiHelpers.headlessDelivery.createSitePage(
		site.id,
		getRandomString(),

		// getPageDefinition([fdsFragmentDefinition])

	);

	return pageLayout;
});
```

However, it is failing due to some missing js modules. Need to investigate a bit more

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:playwright

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:playwright

**Comment by juanjofgliferay** [comment]
ci:stop

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:playwright

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js (hunk 1 above)
Comment (chestofwonders) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.js:901]: Just asking, is there any alternative for a `testid` here (for not losing a chance to test accessibility at the same time) ?

File: modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts (hunk 3 above)
Comment (chestofwonders) [reply modules/test/playwright/tests/frontend-data-set-views-web/settings.spec.ts:194]: Absolutely agree, and we could even create different folders for DSM and fragment ones

**Comment by chestofwonders** [comment]
LGTM✨

**Review by chestofwonders — APPROVED** [review]
LGTM! ✨

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:playwright

**Comment by juanjofgliferay** [comment]
No playwright test failures related

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:report:2555618235

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:report:2557168497

**Comment by juanjofgliferay** [comment]
Failures not related with this PR based on https://github.com/liferay-frontend/liferay-portal/pull/4083#issuecomment-2030253431

**Comment by juanjofgliferay** [comment]
ci:forward

**Pull request closed without merging**